packages/std/src/results/subcall.rs

```diff
@@ -133,10 +133,38 @@ pub struct Event {
 }
 
 impl Event {
-    pub fn new(kind: &str, attributes: Vec<Attribute>) -> Self {
+    pub fn new(kind: impl Into<String>, attributes: Vec<Attribute>) -> Self {
         Event {
-            kind: kind.to_string(),
+            kind: kind.into(),
             attributes,
         }
     }
+
+    pub fn empty(kind: impl Into<String>) -> Self {
+        Self::new(kind, vec![])
+    }
+
+    pub fn attr(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
+        self.attributes.push(crate::attr(key.into(), value.into()));
+        self
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::attr;
+
+    #[test]
+    fn event_builder_test() {
+        let event = Event::empty("test").attr("foo", "bar").attr("bar", "baz");
```

Comment:
Nice. We don't use builder interfaces a lot so far but I see the point.

Until now, Event was never created in contracts. So it would still be possible to rename `Event::empty` to just `Event::new` and require explicit `Event` constructions when all attributes are already existing in a list. Then the two options are:

```rust
// options 1 (preferred)
let event = Event::new("test").attr("foo", "bar").attr("bar", "baz");

// option 2 (rare, advanced)
let event = Event {
    kind: "test".to_string(),
    attributes: pre_constructed_elsewhere
};
```

Reply:
I agree here.
Since `kind` and `attributes` are public, the direct constructor (second example) is always available and no need for functions to do that.

The only benefit I can see there is pre-allocating the proper size of vector. But I guess just doing something like `attributes: Vec::with_capacity(10)` in `empty` would cover all normal use cases efficiently.

Reply:
I like it. Yeah, the main reason I added the new `empty` constructor instead of adapting `new` was because I was worried I'd break an API.